src/checkfiles/checkfiles.py

```diff
@@ -0,0 +1,197 @@
+
+
+def check_valid_gzipped_file_format(errors, is_gzipped, file_format):
+    if file_format in ZIP_FILE_FORMAT and not is_gzipped:
+        errors['gzip'] = f'{file_format} file should be gzipped'
+    elif file_format not in ZIP_FILE_FORMAT and is_gzipped:
+        errors['gzip'] = f'{file_format} file should not be gzipped'
+
+
+def check_file_size(errors, file_size, size_in_cloud_storage):
+    if size_in_cloud_storage != file_size:
+        errors['file_size'] = f'submitted file zise {str(file_size)} does not mactch file zise {str(size_in_cloud_storage)} in cloud storage'
+
+
+def check_md5sum(errors, md5sum, etag, file_path, chunk_size=CHUNK_SIZE):
+    logging.info(f'the eTag is {etag}')
+    if etag != md5sum:
+        md5 = hashlib.md5()
+        with open(file_path, 'rb') as local_file:
+            while chunk := local_file.read(chunk_size):
+                md5.update(chunk)
+        calculated_md5sum = md5.hexdigest()
+
+        if md5sum != calculated_md5sum:
+            errors['md5sum'] = f'submitted file md5sum {(md5sum)} does not mactch file md5sum {calculated_md5sum} in cloud storage'
+
+
+def check_content_md5sum(errors, file_path, chunk_size=128*6400000):
+    md5 = hashlib.md5()
+    with gzip.open(file_path) as f:
+        while chunk := f.read(chunk_size):
+            md5.update(chunk)
+    content_md5sum = md5.hexdigest()
+    logging.info(f'content md5sum is {content_md5sum}')
+    url = CONTENT_MD5SUM_URL + content_md5sum
+    session = requests.Session()
+    username = os.getenv('ENCODE_ACCESS_KEY')
+    password = os.getenv('ENCODE_CECRET_KEY')
+    session.auth = (username, password)
+    conflict_files = session.get(url).json()['@graph']
+    if conflict_files:
+        accessions = []
+        for file in conflict_files:
+            accessions.append(file['accession'])
+        errors[
+            'content_md5sum'] = f"content md5sum conflicts with content md5sum of existing file(s): {', '.join(accessions)}"
+
+
+def bam_pysam_check(errors, file_path, number_of_reads):
+    try:
+        pysam.quickcheck(file_path)
+        result = pysam.stats(file_path)
+        if 'SN\tis sorted:\t0' in result:
+            errors['bam_error'] = 'the bam file is not sorted'
+        else:
+            samfile = pysam.AlignmentFile(file_path, 'rb')
+            count = samfile.count(until_eof=True)
+            logging.info(f'the number of reads: {count}')
+            if count != number_of_reads:
+                errors['bam_error'] = f'sumbitted number of reads {number_of_reads} does not match number of reads {count} in cloud storage'
+            samfile.close()
+    except pysam.utils.SamtoolsError as e:
+        errors['bam_error'] = f'file is not valid bam file by SamtoolsError: {str(e)}'
+
+
+def bam_generate_bai_file(file_path):
+    pysam.index(file_path)
+
+
+def fastq_check(errors, file_path, number_of_reads, read_length):
+    tmp_file_path = 'temp.fasq.gz'
+    shutil.copyfile(file_path, tmp_file_path)
```

Comment:
Why do we want to copy this file around before operating on it? Remember that in production these files may be huge.

Reply:
Using file for pyfastx directly from the mounted file system using goofys will cause error.  I don't have this problem when working with google fuse and google bucket.